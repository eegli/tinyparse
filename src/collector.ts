import { WithPositionalArgs } from './types';

export class Collector<T> {
  constructor(private readonly defaults: T) {}
  public collectWithPositionals(positionals: string[]): WithPositionalArgs<T> {
    return {
      ...this.collect(),
      _: positionals,
    };
  }

  public collect(): T {
    return this.defaults;
  }
}
