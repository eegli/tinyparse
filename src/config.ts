import {
  AnyGlobal,
  CommandOptionsMap,
  DefaultHandler,
  ErrorHandler,
  FlagOptionsMap,
  FlagValueRecord,
  MetaOptions,
  SubparserOptionsMap,
} from './types';

interface CoreConfig<O extends FlagValueRecord, G extends AnyGlobal> {
  meta: MetaOptions;
  options: FlagOptionsMap;
  commands: CommandOptionsMap<O, G>;
  parsers: SubparserOptionsMap;
}

export interface CommonConfig<O extends FlagValueRecord, G extends AnyGlobal>
  extends CoreConfig<O, G> {
  defaultHandler: DefaultHandler<O, G>;
  globalSetter?: (options: O) => G | Promise<G>;
  errorHandler?: ErrorHandler;
}

export interface HelpPrinterConfig<
  O extends FlagValueRecord,
  G extends AnyGlobal,
> extends Partial<CoreConfig<O, G>> {}
