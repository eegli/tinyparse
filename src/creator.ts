export type ConfigLike = Record<string, string | number | boolean>;

export function argvToObj(args: string[]): ConfigLike {
  return args.reduce((acc, curr, idx, orig) => {
    if (curr.startsWith('--')) {
      const arg = curr.slice(2);
      let argVal: string | number | boolean = orig[idx + 1];
      // Assume boolean flag
      if (!argVal || argVal.startsWith('--')) {
        acc[arg] = true;
        // Assume number
      } else if (/^\d+$/.test(argVal)) {
        acc[arg] = +argVal;
        // Assume string
      } else {
        acc[arg] = argVal;
      }
    }
    return acc;
  }, {} as ConfigLike);
}

export function configFactory<T extends ConfigLike, R extends keyof T = string>(
  base: T,
  ...required: Array<R extends string ? R : never>
) {
  return function (args?: Partial<T> | Array<string>): Promise<T> {
    return new Promise((resolve, reject) => {
      if (!args) {
        return resolve(base);
      }
      const cfmap = new Map(Object.entries(base));

      // [--id, 123, --secret, xyz] --> arg, argVal, arg, argVal

      if (!Array.isArray(args)) {
        Object.entries(args).forEach(([arg, argVal]) => {
          if (cfmap.has(arg)) {
            const configVal = cfmap.get(arg);
            // Convert string to number if valid
            if (typeof configVal === 'number' && /^\d+$/.test(argVal)) {
              cfmap.set(arg, +argVal);
            } else if (typeof configVal === typeof argVal) {
              cfmap.set(arg, argVal);
            } else {
              reject(
                `Invalid type for option "${arg}". Expected ${typeof configVal}, got ${typeof argVal}`
              );
            }
          } else {
            console.warn(`Ignoring unknown option "${arg}"`);
          }
        });
      } else {
        args.forEach((val, idx, orig) => {
          if (val.startsWith('--')) {
            const arg = val.slice(2);
            const key = cfmap.get(arg);
            if (cfmap.has(arg)) {
              let argVal: string | number | boolean = orig[idx + 1];
              // Boolean flag that default to false
              if (typeof key === 'boolean') {
                cfmap.set(arg, true);
              }
              // Convert string input to number
              else if (typeof key === 'number' && /^\d+$/.test(argVal)) {
                cfmap.set(arg, +argVal);
              } else if (typeof key === typeof argVal) {
                cfmap.set(arg, argVal);
              } else {
                reject(
                  `Invalid type for option "--${arg}". Expected ${typeof key}, got ${typeof argVal}`
                );
              }
            } else {
              console.warn(`Ignoring unknown option "--${arg}"`);
            }
          }
        });
      }

      // Check for required properties
      required.forEach((prop) => {
        if (!cfmap.get(prop)) {
          reject(`Missing required config property "${prop}"`);
        }
      });

      return resolve(Object.fromEntries(cfmap) as T);
    });
  };
}
