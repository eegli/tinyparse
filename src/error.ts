export class ValidationError extends Error {
  constructor(err: string) {
    super(err);
    this.name = 'ValidationError';
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}
