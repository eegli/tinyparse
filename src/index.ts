import { displayHelp } from './help';
import { parseObjectLiteral } from './parse';
import { transformArgv, transformOptions } from './transform';
import {
  ParsingOptions,
  PositionalArgs,
  SimpleRecord,
  WithPositionalArgs,
} from './types';

export { ValidationError } from './error';

/**
 * Parser factory function. Returns a parser that is bound to the
 * given default values and options per key.
 *
 * @export
 */
export function createParser<T extends SimpleRecord>(
  defaultValues: T,
  parsingOptions?: ParsingOptions<T>
) {
  const options = transformOptions(parsingOptions);
  const filePathArg = parsingOptions?.filePathArg;

  function parse(input?: Partial<T>): Promise<T>;
  function parse(input?: string[]): Promise<WithPositionalArgs<T>>;
  function parse(input: Partial<T> | string[] = {}): Promise<T> {
    let positionalArgs: PositionalArgs | undefined;
    if (Array.isArray(input)) {
      [input, positionalArgs] = transformArgv<T>({
        argv: input,
        filePathFlag: filePathArg?.longFlag,
        options,
      });
    }
    return parseObjectLiteral({
      defaultValues,
      input,
      options,
      forwardArgs: positionalArgs && { _: positionalArgs },
    });
  }

  return {
    help: function (title?: string): string {
      return displayHelp({ defaultValues, options, title, filePathArg });
    },
    parse,
  };
}
