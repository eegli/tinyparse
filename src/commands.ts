import { CommonConfig } from './config';
import { Parser } from './parser';
import {
  AnyGlobal,
  CommandArgPattern,
  DefaultHandler,
  DowncastFlag,
  ErrorHandler,
  FlagOptions,
  FlagOptionsExt,
  FlagValue,
  FlagValueRecord,
  MetaOptions,
  Subcommand,
  Subparser,
} from './types/internals';

export class CommandBuilder<Options, Globals> {
  #config: CommonConfig = {
    meta: {},
    options: new Map(),
    commands: new Map(),
    parsers: new Map(),
    defaultHandler: () => {},
  };

  #takenFlags = new Set<string>();
  #takenCommands = new Set<string>();

  #validateCommand = (command: string): void => {
    if (this.#takenCommands.has(command)) {
      throw new Error(`Command "${command}" has been declared twice`);
    }
  };

  #validateFlag = (flag: string): void => {
    if (this.#takenFlags.has(flag)) {
      throw new Error(`Flag "${flag}" has been declared twice`);
    }
  };

  #validateOption = (key: string): void => {
    if (this.#config.options.has(key)) {
      throw new Error(`Option "${key}" has been declared twice`);
    }
  };

  #validateOneOf = <T>(id: string, type: T, ...options: T[]): void => {
    const expectedType = typeof type;
    const wrongType = options.find((opt) => typeof opt !== expectedType);
    if (wrongType) {
      throw new Error(
        `OneOf for option "${id}" contains invalid type ${typeof wrongType}, expected ${expectedType}`,
      );
    }
  };

  #tryRegisterCommandToken = (command?: string) => {
    if (command) {
      this.#validateCommand(command);
      this.#takenCommands.add(command);
    }
  };

  #tryRegisterFlagToken = (flag?: string) => {
    if (flag) {
      this.#validateFlag(flag);
      this.#takenFlags.add(flag);
    }
  };

  /**
   * Add an option (flag)
   */
  option<
    K extends string,
    V extends FlagValue,
    R extends boolean,
    O extends unknown[],
    Choices = O[number],
    Values = R extends true ? Choices : V | Choices,
  >(
    key: K,
    opts: FlagOptionsExt<V, R, Choices>,
  ): CommandBuilder<Options & Record<K, Values>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ): CommandBuilder<Options & Record<K, DowncastFlag<V>>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ) {
    const { longFlag, shortFlag, oneOf, defaultValue } = opts;

    if (oneOf) {
      this.#validateOneOf(key, defaultValue, ...oneOf);
    }

    this.#validateOption(key);
    this.#config.options.set(key, opts);

    this.#tryRegisterFlagToken(longFlag);
    this.#tryRegisterFlagToken(shortFlag);

    return this as CommandBuilder<
      Options & Record<K, DowncastFlag<V>>,
      Globals
    >;
  }

  /**
   * Add a subcommand
   */
  subcommand<A extends CommandArgPattern>(
    command: string,
    opts: Subcommand<Options, Globals, A>,
  ) {
    this.#tryRegisterCommandToken(command);
    this.#config.commands.set(
      command,
      opts as Subcommand<FlagValueRecord, AnyGlobal, A>,
    );
    return this;
  }

  /**
   * Add a subparser
   */
  subparser(command: string, opts: Subparser) {
    this.#tryRegisterCommandToken(command);
    this.#config.parsers.set(command, opts);
    return this;
  }

  /**
   * Set the globals
   */
  setGlobals<G extends AnyGlobal>(
    setGlobals: (options: Options) => G | Promise<G>,
  ) {
    this.#config.globalSetter = setGlobals as (
      options: FlagValueRecord,
    ) => AnyGlobal | Promise<AnyGlobal>;
    return this as CommandBuilder<Options, Globals & G>;
  }

  /**
   * Set metadata
   */
  setMeta(meta: MetaOptions) {
    this.#tryRegisterCommandToken(meta.help?.command);
    this.#tryRegisterCommandToken(meta.version?.command);

    this.#tryRegisterFlagToken(meta.help?.longFlag);
    this.#tryRegisterFlagToken(meta.help?.shortFlag);
    this.#tryRegisterFlagToken(meta.version?.longFlag);
    this.#tryRegisterFlagToken(meta.version?.shortFlag);

    this.#config.meta = meta;
    return this;
  }

  /**
   * Set the error handler
   */
  onError(handler: ErrorHandler) {
    this.#config.errorHandler = handler;
    return this;
  }

  /**
   * Set the default handler
   */
  defaultHandler(handler?: DefaultHandler<Options, Globals>) {
    return new Parser<Options>({
      ...this.#config,
      defaultHandler:
        (handler as DefaultHandler) || this.#config.defaultHandler,
    });
  }
}
