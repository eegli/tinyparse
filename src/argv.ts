import { ValidationError } from './error';
import { CommandOptions } from './types';
import Utils from './utils';

export class ArgvTransformer {
  public static transform(
    argv: string[],
  ): [Map<string, string | boolean>, string[]] {
    const flagMap = new Map<string, string | boolean>();

    const positionals: string[] = [];
    let isPositional = true;

    for (let i = 0; i < argv.length; i++) {
      const arg = argv[i];

      // A short flag is also a long flag
      const isShortFlag = Utils.isShortFlag(arg);

      if (!isShortFlag && isPositional) {
        positionals.push(arg);
        // Collect positionals
        continue;
      }

      isPositional = false;

      if (!isShortFlag) continue;

      const splitted = Utils.splitAtFirst(arg, '=');

      const [flag] = splitted;
      let [, flagVal] = splitted;

      if (!flagVal) flagVal = argv[i + 1];

      // Assume boolean flag
      if (!flagVal || Utils.isShortFlag(flagVal)) {
        flagMap.set(flag, true);
        // Assume string or number
      } else {
        flagMap.set(flag, flagVal);
      }
    }

    return [flagMap, positionals];
  }

  public static validateCommands(
    positionals: string[],
    commandOptions: CommandOptions,
  ): void {
    const availableCommands = Object.keys(commandOptions);

    // No commands are specified in the config
    if (availableCommands.length === 0) return;

    const [command, ...args] = positionals;

    // User did not specify a command
    if (!command || !commandOptions[command]) return;

    const { args: expectedPattern } = commandOptions[command];

    // Allow any number of command arguments
    if (typeof expectedPattern === 'string') return;

    // Number of arguments is satisfied
    if (expectedPattern.length === args.length) return;

    const tooFewArgs = expectedPattern.length > args.length;

    if (tooFewArgs) {
      throw new ValidationError(
        `Invalid usage of command '${command}'. Too few arguments`,
      );
    }
    throw new ValidationError(
      `Invalid usage of command '${command}'. Too many arguments`,
    );
  }
}
