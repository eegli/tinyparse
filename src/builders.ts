import { transformArgv } from './argv';
import { validateCommandArgs } from './commands';
import { collect } from './flags';
import {
  Subcommand,
  Downcast,
  FlagOptionMap,
  FlagRecord,
  FlagOptions,
  FlagArgValue,
  CommandOptionMap,
  CommandArgPattern,
} from './types';

export class CommandBuilder<F extends FlagRecord = Record<never, never>> {
  #flags: FlagOptionMap = new Map();
  #commands: CommandOptionMap<F> = new Map();

  public subcommand<P extends CommandArgPattern>(
    command: string,
    opts: Subcommand<F, P>,
  ) {
    this.#commands.set(command, opts);
    return this;
  }

  public flag<T extends string, V extends FlagArgValue>(
    flag: T,
    opts: FlagOptions<V>,
  ) {
    this.#flags.set(flag, opts);
    // TODO: Figure out how to make this typecheck properly
    return this as unknown as CommandBuilder<F & Record<T, Downcast<V>>>;
  }

  public build() {
    return new DispatchBuilder<F>(this.#flags, this.#commands);
  }
}

export class DispatchBuilder<F extends FlagRecord> {
  #handler?: (flags: F, positionals: string[]) => void;
  #flags: FlagOptionMap;
  #commands: CommandOptionMap<F>;

  constructor(flags: FlagOptionMap, commands: CommandOptionMap<F>) {
    this.#flags = flags;
    this.#commands = commands;
  }

  public default<T extends F>(
    handler: T extends FlagRecord
      ? (flags: T, positionals: string[]) => void
      : never,
  ) {
    this.#handler = handler;
    return this;
  }

  public parse(argv: string[]): {
    flags: F;
    call: () => void;
  } {
    const [flagMap, positionals] = transformArgv(argv);
    const flags = collect(flagMap, this.#flags) as F;

    const [command, ...args] = positionals;
    const commandOpts = this.#commands.get(command);

    if (command && commandOpts) {
      validateCommandArgs(command, commandOpts, args);
      const call = commandOpts.handler.bind(this, flags, args);
      return {
        flags,
        call,
      };
    }
    if (this.#handler) {
      const call = this.#handler?.bind(this, flags, positionals);
      return {
        flags,
        call,
      };
    }
    return {
      flags,
      call: () => {},
    };
  }
}
