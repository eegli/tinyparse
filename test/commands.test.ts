import { ArgvTransformer } from '../src/argv';

describe('Positionals validation, ok', () => {
  const options: [
    ...Parameters<typeof ArgvTransformer.validateCommands>,
    string,
  ][] = [
    [[], {}, 'default'],
    [['hello-world'], {}, 'no command options'],
    [
      ['x'],
      {
        cd: {
          args: '',
        },
      },
      'unknown command arg',
    ],
    [
      ['ls'],
      {
        ls: {
          args: '',
        },
      },
      'any command arguments 1',
    ],
    [
      ['ls', 'folder1', 'folder2'],
      {
        ls: {
          args: '',
        },
      },
      'any command arguments 2',
    ],
    [
      ['info'],
      {
        info: {
          args: [],
        },
      },
      'no command arguments',
    ],
    [
      ['cp', 'folder1', 'folder2'],
      {
        cp: {
          args: ['', ''],
        },
      },
      'multiple command argument',
    ],
  ];

  for (const [positionals, opts, desc] of options) {
    test(`validation, ${desc}`, () => {
      expect(() =>
        ArgvTransformer.validateCommands(positionals, opts),
      ).not.toThrow();
    });
  }
});

describe('Positionals validation, fail', () => {
  const options: [
    ...Parameters<typeof ArgvTransformer.validateCommands>,
    string,
  ][] = [
    [
      ['cd'],
      {
        cd: {
          args: ['folder'],
        },
      },
      'too few command args',
    ],
    [
      ['cd', 'folder1', 'folder2'],
      {
        cd: {
          args: [''],
        },
      },
      'too many command args',
    ],
  ];

  for (const [positionals, opts, desc] of options) {
    test(`validation, ${desc}`, () => {
      expect(() =>
        ArgvTransformer.validateCommands(positionals, opts),
      ).toThrowErrorMatchingSnapshot();
    });
  }
});
