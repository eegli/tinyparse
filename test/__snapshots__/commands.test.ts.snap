// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Positionals validation, fail validation, too few command args 1`] = `"Invalid usage of command 'cd'. Too few arguments"`;

exports[`Positionals validation, fail validation, too many command args 1`] = `"Invalid usage of command 'cd'. Too many arguments"`;
